File: 146_lru_cache/LRUCache.java
public class LRUCache {
  private class Node {
    int key, value;
    Node prev, next;

    public Node() {
      key = 0;
      value = 0;
    }

    public Node(int k, int v) {
      key = k;
      value = v;
    }
  }

  private int capacity, count;
  private Map<Integer, Node> map;
  private Node head, tail;

  public LRUCache(int capacity) {
    this.count = 0;
    this.capacity = capacity;
    this.map = new HashMap<>(); 
    head = new Node();
    tail = new Node();
    head.next = tail;
    tail.prev = head;
  }

  public int get(int key) {
    Node node = map.get(key);
    if(null == node) {
      return -1;
    }
    remove(node);
    addAtHead(node);
    return node.value;
  }

  public void set(int key, int value) {
    Node node = map.get(key);
    if(null == node) {
      if(count >= capacity) {
        Node nodeToRemove = tail.prev;
        nodeToRemove.prev.next = nodeToRemove.next;
        nodeToRemove.next.prev = nodeToRemove.prev;
        nodeToRemove.prev = null;
        nodeToRemove.next = null;
        map.remove(nodeToRemove.key);
      } else {
        count ++;
      }
      node = new Node(key, value);
      addAtHead(node);
      map.put(key, node);
    } else {
      node.value = value;
      remove(node);
      addAtHead(node);
    }
  }

  private void remove(Node node) {
    node.prev.next = node.next;
    node.next.prev = node.prev;
    node.prev = null;
    node.next = null;
  }

  private void addAtHead(Node node) {
    Node after = head.next;
    node.next = after;
    node.prev = head;
    head.next = node;
    after.prev = node;
  }
}

File: 146_lru_cache/lru_cache.rb
class LRUCache
  # Initialize your data structure here
  # @param {Integer} capacity
  def initialize(capacity)
    @list = LinkedHashMap.new(capacity)
  end

  # @param {Integer} key
  # @return {Integer}
  def get(key)
    @list.get(key)
  end

  # @param {Integer} key
  # @param {Integer} value
  # @return {Void}
  def set(key, value)
    @list.set(key, value)
  end
end

class LinkedHashMap
  def initialize(capacity)
    @count     = 0
    @capacity  = capacity
    @map       = {}
    @head      = Node.new
    @tail      = Node.new
    @head.next = @tail
    @tail.prev = @head
  end

  def get(key)
    node = @map[key]
    if node
      refresh(node)
      node.value
    else
      -1
    end
  end

  def set(key, value)
    return if invalid_capacity?
    node = @map[key]
    if node
      node.value = value
      refresh(node)
      return
    end

    node = Node.new(key, value)
    if insufficient_capacity?
      last_node = @tail.prev
      @map.delete(last_node.key)
      remove(last_node)
    end

    shift(node)
    @map[key] = node
  end

  private

  def invalid_capacity?
    @capacity <= 0
  end

  def insufficient_capacity?
    @count >= @capacity
  end

  def refresh(node)
    remove(node)
    shift(node)
  end

  def shift(node)
    @count    += 1
    first      = @head.next
    @head.next = node
    node.prev  = @head
    node.next  = first
    first.prev = node
  end

  def remove(node)
    @count        -= 1
    prev_node      = node.prev
    next_node      = node.next
    prev_node.next = next_node
    next_node.prev = prev_node
    node.prev = node.next = nil
  end

end

class Node
  attr_accessor :key, :value, :prev, :next
  def initialize(key = nil, value = nil)
    @key   = key || 0
    @value = value || 0
  end
end